### DISCUSSION TITLE
Inject object returned by suspend function

### DISCUSSION THREAD

**Discussion by ivanovpv** [InsertKoinIO/koin#1923] [General]
have object which created by suspend function like:

`suspend fun someSuspendFunction(): MyPreciousObject`

And I need to provide it for later injection using koin, like:

```
val myModule = module {
       single<MyPreciousObject> {
         //can call suspend function only from a coroutine or another suspend function
         someSuspendFunction() 
    }
}

```

Any ideas how to resolve this kind of situation? And please bear in mind that I can't change `someSuspendFunction()`

P.S. I'm using Kotlin JS/WASM environment, so `runBlocking` doesn't exist
P.P.S. Original [question](https://stackoverflow.com/questions/78747425/koin-inject-object-returned-by-suspend-function) - in StackOverflow

**Comment by arnaudgiuliani** [comment]
for now 3.5/4.0 you can't do such definition. Things need to challenged in 4.x

**Comment by ragboyjr** [comment]
I've come across this type of problem when I want to load secrets from aws secrets manager client and have the loaded secrets be registered as an instance in koin.

I basically just solve this with mutable variables and doing the suspending / loading outside of koin. a trivial example looks like:

```kotlin
class SecretsStore {
  var secrets: Secrets? = null

  suspend fun load() {
    // do the work
    // set secrets
  }
}

data class Secrets(val password: String)

val koinApp = koinApplication {
  modules(module {
    single<SecretsStore> { SecretsStore() }
    single<Secrets> {
      checkNotNull(get<SecretsStore>().secrets)
    }
  })
}

// do this early on in the app lifecycle before resolving objects that depend on Secrets
koinApp.get<SecretsStore>().load()
```
